### ISSUE TITLE
Switching between navigators via conditional rendering can corrupt navigation state

### ISSUE THREAD

**Issue by TheAlmightyBob** [react-navigation/react-navigation#9436]
Sorry for the confusing title, but this is a bit complex.

(also, this makes extensive mention of tab navigators, but I did not file the issue in that category because the root issue seems to be in React Navigation's core state management, not any of the actual tab navigator code)

**Current Behavior**

If you have a stack navigator, and a tab navigator with nested stack navigators, and simply switch between the two with conditional rendering (rather than navigating via some parent navigator), the navigation state can become corrupted such that switching to the tab navigator resurrects its previous navigation state rather than initializing new state and loading the initial route.

I've created a simple repro here, I think it's probably easier to play with than to watch a video: https://snack.expo.io/L-5L-A411

The example use case is when switching between "authenticated"/"unauthenticated" navigators, as in https://reactnavigation.org/docs/auth-flow (but since different types of navigators are used, the entire navigators are swapped out rather than just swapping screens within a Stack Navigator).

In this repro:
 * If you log in, switch to Tab 2, log out, then log back in again, you are back to Tab 1. (This seems normal and desired.)
 * But if you log in, push a new screen on either Tab 1 or Tab 2, switch to Tab 2, log out, then log back in again, you are still left on Tab 2! (though reset to the root of the stacks)
 * If you _then_ log out and back in without pushing a new screen, you will now be on Tab 1 as intended.

What is happening in the React Navigation code:
 * When switching from the Tab Navigator to the Welcome Stack, if any of the child stack navigators of the Tab Navigator have performed any navigation, the unmount handler in `useNavigationBuilder` calls `setState(undefined)`
   * (if no stack navigation occurred, this does not happen and the bug does not repro)
 * That "setState" is actually `SceneView`'s `setCurrentState`, which first calls `useNavigationBuilder`'s `getState`, which calls `useSyncState`'s `getState`, which returns the correct new state for the `WelcomeStack`.
 * `useNavigationBuilder`'s `getState` then checks `isStateInitialized`, which actually returns `false` (!), because the type of the current state ("stack") does not match the type of the router's state ("tab").
   *  presumably here what's happening is that this child stack navigator of the tab navigator thinks that the tab navigator must still be around and disregards any state that contradicts that belief?
 * Since `isStateInitialized` returned `false`, `useNavigationBuilder`'s `getState` ignores the current state (which was correct) and instead returns the outdated `initializedStateRef`, which is still a copy of the old Tab Navigator state. (not clear to me how `initializedStateRef` is managed, whether that's expected behavior or not...)
 * This outdated state is now what `SceneView`'s `setCurrentState` passes to `useSyncState`'s `setState`, which _replaces the correct state_ (that of the `WelcomeStack`) _with the wrong state_ (that of the now-removed Tab Navigator).
 * Now, if you press "Log In" again to switch back to the Tab Navigator, the creation of the Tab Navigator sees that there is already a tab state so it just uses that rather than initializing a new state.
   *  (if some additional stack navigation was first performed within the `WelcomeStack`, the state would be corrected and the bug would no longer repro)

I don't know the normal state management code well enough to know where exactly things go wrong/unexpected... it _seems_ like the core problem is that unmounting nested navigators when their parent navigator has gone away causes them to set the state back to that of the parent navigator..... should they have done their cleanup in advance? Should they just skip cleanup if hosted in a navigator that is going/gone away?

**Expected Behavior**

I expect that switching to the Tab Navigator via conditional rendering should consistently reinitialize it each time and show the initial route.

Moreso, I expect that the React Navigation state should be up-to-date, not continue to reflect the state of an old unmounted navigator instead of the current navigator...

**How to reproduce**

https://snack.expo.io/L-5L-A411 (repro steps listed above)

**Your Environment**

| software                       | version |
| ------------------------------ | ------- |
| iOS or Android                 | both, and web
| @react-navigation/native       |  5.9.3
| react-native                   |  0.62.2
| node                           | 15.11.0
| npm or yarn                    |  yarn 1.22.10

**Comment by karlhorky** [comment]
Yeah, the conditional rendering pattern below lead to a `Warning: Can't perform a React state update on an unmounted component` error message today with @thomalexg:

```js
<NavigationContainer>
  {user ? <AppNavigator /> : <AuthNavigator setUser={setUser} />}
</NavigationContainer>
```

Where `AppNavigator` is:

```js
<Tab.Navigator>
  <Tab.Screen /* ... */ />
  <Tab.Screen /* ... */ />
</Tab.Navigator>
```

and `AuthNavigator` is:

```js
<Stack.Navigator>
  <Stack.Screen /* ... */ />
  <Stack.Screen /* ... */ />
</Stack.Navigator>
```

---

Full error:

```
Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in %s.%s, a useEffect cleanup function,
```

**Comment by raajnadar** [comment]
If you have any async api call and if you switch nav state it will surely cause a memory leak. You need to unsubscribe to async task and then navigate.

**Comment by karlhorky** [comment]
Ok, so what you're saying is that the code in my comment above is actually an antipattern for React Navigation? Can you show a code example with a better approach?

**Comment by raajnadar** [comment]
Either in `<AppNavigator />` or `<AuthNavigator setUser={setUser} />` there should not be a async task running while you change the value of var `user`.

Eg: - Lets say you have a async task running in the `<AuthNavigator setUser={setUser} />` and you populate some value in the `users` var it will switch the nav and render `<AppNavigator />`. once the async task finishes it will search for the component which started but due to the ternary operation that component will not be available.

`Warning: Can't perform a React state update on an unmounted component`

That causes the above warning.

**Comment by TheAlmightyBob** [comment]
Yeah I agree with @raajnadar, I don't think the `unmounted component` warning is related to this issue.

**Comment by TheAlmightyBob** [comment]
(another option, if you can't "unsubscribe" the async call, is to keep track of whether the component is mounted and ensure the completion of the async task doesn't try to update the component if it's unmounted.... libraries/patterns like Redux are also useful for avoiding this issue)

**Comment by ctwillie** [comment]
@TheAlmightyBob Did you find any solution/work around for this? I am experiencing this exact issue. And thanks for the work of digging into the issue and describing the buggy behavior so well.

**Comment by TheAlmightyBob** [comment]
@ctwillie The workaround I'm using, though I can't guarantee it'd work for all cases, is to always have a root stack navigator that never changes. It is never actually used for any stack navigation, it just is there to provide some stability to the navigation state. 

e.g., to borrow the sample posted above, instead of
```
<NavigationContainer>
  {user ? <AppNavigator /> : <AuthNavigator setUser={setUser} />}
</NavigationContainer>
```

It would be something like (this is quick pseudocode, don't just copy/paste it):
```
<NavigationContainer>
  <Stack.Navigator>
    {user ? <Stack.Screen component={AppNavigator}/> : <Stack.Screen component={AuthNavigator}/>}
  </Stack.Navigator>
</NavigationContainer>
```

**Comment by ctwillie** [comment]
Ahh, I see. Good solution.

As you stated, I'm not sure if this would work with my use case, but it definitely might. This was my case and solution for anyone it may help.

The Problem:

I have two navigators conditionally rendered by the authenticated user.
```js
<NavigationContainer>
  {authUser ? <AppNavigator /> : <LoginNavigator />}
</NavigationContainer>
```
The specific problem I was facing is when a user was inside AppNavigator and navigated to the LogoutScreen and logged out, which just sets authUser to null, the LoginNavigator would be rendered. So far so good. But, when they reauthenticated, and the AppNavigator is rendered, the LogoutScreen would be shown, (because of the navigation state issue pointed out by @TheAlmightyBob). In my app, this prevented certain necessary data from being set in the apps context.


My Solution:

I noticed that creating another navigation action corrected the navigation state. So instead of just setting the authUser to null on the LogoutScreen, I navigate to a ForceLogoutScreen that immediately logs the authUser out.

```js
export default function ForceLogout() {
  const auth = useAuth();

  React.useEffect(() => {
    auth.logout();
  }, []);

  return <View></View>;
}
```

This way, I was still able to log the user out and make an extra navigation action to correct the broken navigation state. So now, when AppNavigator is rendered it will go to my defined initial route instead of using the stale navigation state.

**Comment by singhsaurabh920** [comment]
`import React, {  } from 'react';
import { useSelector } from 'react-redux';
import { NavigationContainer } from '@react-navigation/native';
import { createNativeStackNavigator } from '@react-navigation/native-stack';
//
import AuthStackNavigator from './AuthStackNavigator';
import MainStackNavigator from './MainStackNavigator';

export default function Routes() {

  const Stack = createNativeStackNavigator();
  const isLogged = useSelector((state) => state.auth.isLogged);


  return (
    <NavigationContainer>
      <Stack.Navigator screenOptions={{ headerShown: false }}>
        {isLogged ? <Stack.Screen name={"MAIN"}  component={MainStackNavigator}/> : <Stack.Screen name={"AUTH"} component={AuthStackNavigator}/>}
      </Stack.Navigator>
    </NavigationContainer>
  );
}`

This is my routes I am surprised its working in android  fine but in ios it crash when i login 


**This is my  MainStackNavigator**

`import * as React from 'react';
import { createNativeStackNavigator } from '@react-navigation/native-stack';
import NavigationStrings from '../constants/NavigationStrings';
import DashboardTabNavigator from './DashboardTabNavigator';
import {
    Contact,
    ChangePassword
} from '../screens';

const MainStack = createNativeStackNavigator();

const MainStackNavigator = () => {
    return (
        <MainStack.Navigator screenOptions={{ headerShown: false }}>
            <MainStack.Screen name={NavigationStrings.DASHBOARD_TAB_NAVIGATOR} component={DashboardTabNavigator} />
            <MainStack.Screen name={NavigationStrings.CONCACT} component={Contact} />
            <MainStack.Screen name={NavigationStrings.CHANGE_PASSWORD} component={ChangePassword} />
        </MainStack.Navigator>
    );
};

export default MainStackNavigator;`

**This is my auth stack**

`import * as React from 'react';
import { View, Text } from 'react-native';
import { NavigationContainer } from '@react-navigation/native';
import { createNativeStackNavigator } from '@react-navigation/native-stack';
import NavigationStrings from '../constants/NavigationStrings';
import DashboardTabNavigator from './DashboardTabNavigator';

import {
  Login,
  LoginV1,
  Contact,
  ForgetPassword,
  ChangePassword,
} from '../screens';

const AuthStack = createNativeStackNavigator();

function AuthStackNavigator() {
  return (
    <AuthStack.Navigator screenOptions={{ headerShown: false }}>
      <AuthStack.Screen name={NavigationStrings.LOGIN} component={Login} />
      <AuthStack.Screen name={NavigationStrings.LOGINV1} component={LoginV1} />
      <AuthStack.Screen name={NavigationStrings.CONCACT} component={Contact} />
      <AuthStack.Screen name={NavigationStrings.FORGET_PASSWORD} component={ForgetPassword} />
      <AuthStack.Screen name={NavigationStrings.CHANGE_PASSWORD} component={ChangePassword} />
    </AuthStack.Navigator>
  );
}

export default AuthStackNavigator;`

**Comment by soarb** [comment]
Hi all, further to this, not sure if it's directly related. I have exactly the same setup, a possible 3 screens that I conditionally register with the wrapping Stack navigator. State _seems_ fine when signing in/signing out, but when the screens get swapped out I get the following React "red alert" Dialog pop up (sorry for the formatting):-

`Exception ‘Bridge module (null) does not conform to RCTBridgeModule’ was thrown while invoking dispatchViewManagerCommand on target UIManager with params (
    3755,
    reset,
        (
    )
)
callstack: (
	0   CoreFoundation                      0x00000001b55265ec CEB74A0C-5EA4-3F5A-9DE8-73F0DAD2F4CB + 972268
	1   libobjc.A.dylib                     0x00000001ad847c00 objc_exception_throw + 60
	2   Foundation                          0x00000001b4a904d4 60F744F4-1345-325E-8970-D37BB5A1A31D [+ 7062740](tel:+7062740)
	3   XXXXXX                              0x0000000102f83ea8 RCTBridgeModuleNameForClass + 220
	4   XXXXXX                              0x0000000103058b28 -[RCTUIManager dispatchViewManagerCommand:commandID:commandArgs:] + 780
	5   CoreFoundation                      0x00000001b546b134 CEB74A0C-5EA4-3F5A-9DE8-73F0DAD2F4CB + 205108
	6   CoreFoundation                      0x00000001b546abcc CEB74A0C-5EA4-3F5A-9DE8-73F0DAD2F4CB + 203724
	7   CoreFoundation                      0x00000001b546a9f4 CEB74A0C-5EA4-3F5A-9DE8-73F0DAD2F4CB + 203252
	8   XXXXXX                              0x0000000103005edc -[RCTModuleMethod invokeWithBridge:module:arguments:] + 1808
	9   XXXXXX                              0x0000000103009e0c _ZN8facebook5reactL11invokeInnerEP9RCTBridgeP13RCTModuleDatajRKN5folly7dynamicEiN12_GLOBAL__N_117SchedulingContextE + 1688
	10  XXXXXX                              0x00000001030095a8 _ZZN8facebook5react15RCTNativeModule6invokeEjON5folly7dynamicEiENK3$_0clEv + 128
	11  XXXXXX                              0x000000010300951c ___ZN8facebook5react15RCTNativeModule6invokeEjON5folly7dynamicEi_block_invoke + 28
	12  libdispatch.dylib                   0x00000001bd3656a8 149FF9CD-3E92-321D-B92F-EF58717C70B6 + 9896
	13  libdispatch.dylib                   0x00000001bd367300 149FF9CD-3E92-321D-B92F-EF58717C70B6 + 17152
	14  libdispatch.dylib                   0x00000001bd36e894 149FF9CD-3E92-321D-B92F-EF58717C70B6 + 47252
	15  libdispatch.dylib                   0x00000001bd36f3c4 149FF9CD-3E92-321D-B92F-EF58717C70B6 + 50116
	16  libdispatch.dylib                   0x00000001bd37a004 149FF9CD-3E92-321D-B92F-EF58717C70B6 + 94212
	17  libdispatch.dylib                   0x00000001bd379878 149FF9CD-3E92-321D-B92F-EF58717C70B6 + 92280
	18  libsystem_pthread.dylib             0x000000021e626964 _pthread_wqthread + 288
	19  libsystem_pthread.dylib             0x000000021e626a04 start_wqthread + 8
)
RCTFatal
facebook::react::invokeInner(RCTBridge*, RCTModuleData*, unsigned int, folly::dynamic const&, int, (anonymous namespace)::SchedulingContext)
facebook::react::RCTNativeModule::invoke(unsigned int, folly::dynamic&&, int)::$_0::operator()() const
invocation function for block in facebook::react::RCTNativeModule::invoke(unsigned int, folly::dynamic&&, int)
149FF9CD-3E92-321D-B92F-EF58717C70B6
149FF9CD-3E92-321D-B92F-EF58717C70B6
149FF9CD-3E92-321D-B92F-EF58717C70B6
149FF9CD-3E92-321D-B92F-EF58717C70B6
149FF9CD-3E92-321D-B92F-EF58717C70B6
149FF9CD-3E92-321D-B92F-EF58717C70B6
_pthread_wqthread
start_wqthread (edited)`

This has been built on XCode 15 and deployed onto a device running iOS 17.

Release builds are fortunately working fine, but the error is a little alarming when developing.

Am using (I think) the latest relevant deps:-
`
    "@react-navigation/drawer": "6.6.4",
    "@react-navigation/native": "6.1.8",
    "@react-navigation/stack": "6.3.18",
`

If I do the same build using XCode 14 onto a pre ios 17 device I don't get the react error in dev mode.

Not sure if this is something anyone is aware of with XCode 15 and iOS 17?

Cheers

Benj

**Comment by pj-prize** [comment]
Yeah, I'm having the same issue as you, @singhsaurabh920 .  I've got some conditionally rendered stacks that work just fine on Android but, even though the boolean value is changing, nothing seems to happen on iOS.  Interestingly enough, I've even wrapped the rendering of stacks into its own function and logged the output and the output is correct!  It's like the old stack is being cached on iOS but not on Android?  There are no errors or anything.  Weird behavior.  Help here would be greatly appreciated!

**Comment by hosein-moayedi** [comment]
@singhsaurabh920 @pj-prize 

Yes, me too.
this is my code and when I am changing my condition with redux, I will see just white screen and nothing (just in IOS not in android)!!

I very tried everything but not working again.

I hope that developers fix it 🌹

**Comment by hosein-moayedi** [comment]
@singhsaurabh920 @pj-prize
I think found the solution:
I think when you are defining your screens in root stack, in IOS, your screens will create and if you using from conditions for render your screen that are unmuted that is not working 

So, my solution is using from separate navigator for root navigation container like this:

const RootNavigation = () => {
  const { authorized } = useSelector((store: RootState) => store.user)

  return (
    <NavigationContainer>
     {authorized ? <AuthorizedStack /> : <NonAuthorizedStack />}
    </NavigationContainer>
  );
};

-------------------------------

const AuthorizedScreensStack = createNativeStackNavigator();

const AuthorizedStack = () => {
  return (
    <AuthorizedScreensStack.Navigator>
      <AuthorizedScreensStack.Screen
        name={'home'}
        component={HomeScreen.component}
        options={{ headerShown: false }}
      />
    </AuthorizedScreensStack.Navigator>
  );
};

-------------------------------

const NonAuthorizedScreensStack = createNativeStackNavigator();

const NonAuthorizedStack = () => {
  return (
    <NonAuthorizedScreensStack.Navigator>
      <NonAuthorizedScreensStack.Screen
        name={'login'}
        component={LoginScreen.component}
        options={{ headerShown: false }}
      />
    </NonAuthorizedScreensStack.Navigator>
  );
};

**Comment by maksymhcode-care** [comment]
Using @react-navigation/stack in combination with @react-navigation/native-stack worked for me
@react-navigation/native-stack does not work with conditional screens, @react-navigation/stack works just fine. Combination can help if using @react-navigation/native-stack is required
```
const Stack = createNativeStackNavigator();
const NStack = createStackNavigator();

    <NavigationContainer
      >
        {!token ? (
          <StackN.Navigator>
            <StackN.Screen
              name={LogInScreen}
              component={LogInScreen}
            />
          </StackN.Navigator>
        ) : (
          <Stack.Navigator>
            <Stack.Screen
              name={MyScreen}
              component={MyComponent}
            />
          </Stack.Navigator>
        )}
      </NavigationContainer>
```
